docs/versioned_docs/version-zenith/quickstart/822194-daggerize.mdx

```diff
@@ -0,0 +1,246 @@
+
+```shell
+dagger install github.com/shykes/daggerverse/hello@v0.1.2
+```
+
+Once the command completes, look at the `dagger.json` file again. You should see that a new dependency has been added, similar to that shown below:
+
+```json
+{
+  "name": "greetings-api",
+  "dependencies": [
+    {
+      "name": "hello",
+      "source": "github.com/shykes/daggerverse/hello@7884cbc1941f85a7c714fc7f1bcc5c148842c21e"
+    }
+  ],
+  "engineVersion": "v0.9.11"
+}
+```
+
+:::note
+The exact Git commit for the module version is recorded in `dagger.json`. Dagger enforces "version pinning", which guarantees that the module version you install is the one you'll be using.
+:::
+
+Go ahead and commit this change as well:
+
+```shell
+git add dagger.json
+git commit -m 'dagger install github.com/shykes/daggerverse/hello'
+```
+
+### Use the installed module
+
+4. Now that your module has its first dependency installed, you can start using it! To do so, simply refer to it by its short name - in this case, `hello`.
+
+List available functions:
+
+```shell
+dagger -m hello functions
+```
+
+You should see a familiar result:
+
+```shell
+Name    Description
+hello   Say hello to the world!
+```
+
+Call a function:
+
+```shell
+dagger -m hello call hello
+```
+
+Again, the output should be familiar:
+
+```shell
+hello, world!
+```
+
+### Daggerize a Go build
+
+The `hello` module is a good example to start with, but it doesn't have any functions to work with a Go project. To Daggerize the Go build for this example project, install the Go builder module from a previous section as another dependency:
+
+```shell
+dagger install github.com/kpenfound/dagger-modules/golang@v0.1.5
+```
+
+Then, call the `Build()` function from the Go builder module to build the project and chain other function calls to export the built binary to the host:
+
+```shell
+dagger -m golang call \
+  # call the go builder function
+  build --project . --os=$(uname -s) --arch=$(uname -m) --args . \
+  # select the binary file produced by the build
+  file --path=./greetings-api \
+  # export
+  export --path=./api-server
+```
+
+Test the newly-built binary by starting the service and sending it a request:
+
+```shell
+./api-server &
+curl localhost:8080
+```
+
+You should see the following output:
+
+```shell
+got / request from 127.0.0.1:51418
+{"greeting":"Hello Daggernauts!"}
+```
+
+At this point, you've partly Daggerized your project, to the extent of building a binary with Dagger. Next, how about containerizing and publishing it?
+
+An easy way to do this is to first add the Wolfi container builder module from earlier as another dependency:
+
+```shell
+dagger install github.com/shykes/daggerverse/wolfi@v0.1.2
+```
+
+Then, add the built binary from the previous step to a Wolfi base container image, and publish it:
+
+```shell
+dagger -m wolfi call \
+  # build a Wolfi container image
+  # add the binary file
+  container with-file --path=/usr/local/bin/api-server --source=./api-server \
```

Comment:
cc @kpenfound @vikram-dagger 

I think this example will raise eyebrows. If I run the quickstart on a Mac, this will publish a container with a Mac binary. So in addition to raising eyebrows, it will break for many users.

I'm not sure what's the simplest way to fix this, without breaking the editorial flow of the quickstart.

Reply:
I think there's 3 options, not sure which is least bad
- do the build again, but with `--os linux`
- only build with `--os linux` and don't run the binary
- update the build function to build multiple os/arch targets and return as a directory

Reply:
Checking to be sure: is this the same as the `uname` issue [briefly discussed above](https://github.com/dagger/dagger/pull/6707#discussion_r1499586134)? 

If yes, one option may be to state the platform upfront ie. "...call the `Build()` function...to build a Go binary for the `linux/amd64` platform..." and specify the arguments inline instead of retrieving via `uname`. Because even on a Linux cloud host, `uname` values don't exactly match what Go expects.

Reply:
Any chance we could find a way to *not* involve the local filesystem at all when building a container? In addition to not being realistic (for a real build you would just chain everything together directly, without export), in the context of a quickstart, we will be sending beginners down the wrong path... There has already been confusion in the past about the role of `export` and the local filesystem. This would compound such confusion.

Since we can't "stitch" calls together by ID in the CLI yet (soon I hope!). Maybe we could add a function in your module @kpenfound that just builds straight to container? WDYT?

Reply:
@shykes that makes sense. I could add a light weight / copy paste `build-container` function that returns a container instead of a file. Alternatively I could refactor it like we talked about at one point with custom types, so that `Build()` returns a custom type from which you can query a file or container. Something like `dagger call build --project foo container publish ttl.sh/foo` + `dagger call build --project foo binary export -o ./foo`

Reply:
I actually had a PR I meant to open, with a `build-container` function :) https://github.com/kpenfound/dagger-modules/pull/2

Reply:
Oh perfect, I'll merge that

Reply:
Released as 0.1.6 https://daggerverse.dev/mod/github.com/kpenfound/dagger-modules/golang@88795761b618360c9d0977a5f2edab77d00c3987

Going through and cleaning up that module with docs/comments now

Reply:
@shykes since we now have this `build-container` function, I guess we don't need the wolfi module for this section?

If correct, then can you suggest a second module to use in this and the next section? Because ideally in the next section, we want to demonstrate connecting two modules with code. 

One idea is to use golang+ttlsh...we could do this, although it isn't a very good example because there is already a built-in `publish`. 

Another option: scan the built container with [trivy](https://daggerverse.dev/mod/github.com/jpadams/daggerverse/trivy@68cf0b4e8998332b5eb0fc63a007847d1e3b538c)?

Reply:
> @shykes since we now have this `build-container` function, I guess we don't need the wolfi module for this section?

I think that's right. If there was a reliable way to stitch calls together from the CLI, then we could do that. But that's not available yet. We can revisit when that feature is available.

> If correct, then can you suggest a second module to use in this and the next section? Because ideally in the next section, we want to demonstrate connecting two modules with code.

My original plan was to use the optional `base` argument to @kpenfound 's `build-container` function, and pass a custom base container, using the wolfi module. That seemed realistic enough, and shows the power of composition with code.

WDYT?

Reply:
Added example per above